API Token Scopes (#145)
val-town/val-town-blog: src/content/blog/api-token-scopes.md

---
title: "Leveling Up Our API Tokens: Introducing Scopes"
description: Improving security and allowing more granular control over permissions.
pubDate: Oct 3, 2024
author: Max McDonnell
---

We're adding more granular permissions to our API tokens and changing some of

> the default behavior of our platform to keep things more secure.

Reviewer: ```suggestion
the default behavior of our platform to make things more secure.
```

should either be "make things more secure" or "keep things secure"

We're introducing read and read/write scopes for five key resources: `val`,
`blob`, `user`, `sqlite`, `email`. You can browse our API docs to see how these
resources are accessed today: https://docs.val.town/openapi.

## Configuring scopes

There will be two ways to configure token scopes. You can visit the
[API tokens](https://www.val.town/settings/api) page and generate a new token
with specific permissions. This token can be used with our API from outside of
the platform, or from within a running Val.

Additionally, you'll be able to configure the permissions of each Val in your
account to ensure that it only has access to the resources that it needs.

## Safer Defaults

> Vals will now use short-lived API tokens when they are running. Additionally,
> the default permissions for a val will only be able to read the `val` resource,

Reviewer: ```suggestion
Vals will now use short-lived API tokens when they are running.,
The default permissions for a val will only be able to read the `val` resource,
```

Too much Additionally in these consecutive sentences imho

not edit, create, or update any Val.

> These new defaults help limit potential damage from misconfigurations,

Reviewer: ```suggestion
These new defaults help limit potential damage from misconfiguration,
```

I think misconfiguration can be treated as a collective noun here.

vulnerable libraries, or account compromises. Without the ability to mutate Val
resources, unauthorized actors cannot execute new code or modify existing code
in your account.

## Migrating current users

> On **November 1st** we'll update the existing API tokens on people's accounts to

Reviewer: ```suggestion
On **November 1st 2024** we'll update the existing API tokens on people's accounts to
```

use these safer defaults. Many users use their existing API tokens or Vals to
create/edit/delete existing Vals in their accounts. We've identified those

> users, We'll hold off on migrating their tokens over to the new permissions, and

Reviewer: ```suggestion
users: we'll hold off on migrating their tokens over to the new permissions, and
```

---
title: "API Tokens Scopes"
description: Improving security with granular control over permissions
pubDate: Oct 3, 2024
author: Max McDonnell
---

> We've added scopes to Val Town API tokens. This will give you more granular
> control over what your API tokens can do.

Reviewer: ```suggestion
We've added scopes to Val Town API tokens to give you more granular control over
what vals can do and more security for your custom uses of the Val Town [REST API](https://docs.val.town/openapi).
```

- `sqlite` – [sqlite database](https://docs.val.town/openapi#tag/sqlite)
- `email` – [ability to send emails](https://docs.val.town/openapi#tag/emails)

You will be able to configure the scopes of your API tokens from the
[API tokens](https://www.val.town/settings/api) page.

## Scoping vals

Vals run with a `valtown` environment variable. You can now configure the scopes
that each of your vals has access to, ensuring that it only has access to the
resources that it needs.

![](./api-token-scopes/val-names-permissions.png)

Vals also use short-lived API tokens when they are running, limiting the
potential damage from a compromised val.

## Safer Defaults

The new default scopes for a val will include everything except `val:write` and
`user:write` to limit potential damage from misconfigurations, vulnerable
libraries, or account compromises.

Without the ability to mutate Val resources, unauthorized actors cannot execute
new code or modify existing code in your account.

We're also considering even fewer default scopes in the future, and having a
slick UI to let you opt-into extra scopes at runtime when you code accesses
them.

## Github Secret Scanning

> We're joining the Github Secret Scanning program, so if your API token is leaked

Reviewer: ```suggestion
We're joining the Github [Secret Scanning program](https://docs.github.com/en/code-security/secret-scanning/introduction/about-secret-scanning), so if your API token is leaked
```
